File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.cpp
#include <string>
#include <vector>
#include <set>

using namespace std;

set<set<string>> s;
vector<bool> is_used;
vector<string> cUser, cBanned;
int total_user, total_ban;

bool match(string l, string r) {
    if(l.length() != r.length()) return false;
    
    for(int i = 0; i < l.length(); i++) {
        if(l[i] == r[i]) continue;
        else if(r[i] == '*') continue;
        return false;
    }
    return true;
}

void dfs(int banned_idx) {
    if(banned_idx == total_ban) {
        set<string> tmp;
        for(int i = 0; i < total_user; i++) {
            if(is_used[i]) tmp.insert(cUser[i]);
        }
        s.insert(tmp);
        return;
    }
    
    for(int idx = 0; idx < total_user; idx++) {
        if(is_used[idx]) continue;
        if(!match(cUser[idx], cBanned[banned_idx])) continue;
        is_used[idx] = true;
        dfs(banned_idx + 1);
        is_used[idx] = false;
    }
}

int solution(vector<string> user_id, vector<string> banned_id) {
    is_used = vector<bool>(user_id.size(), false);
    cUser = user_id; 
    cBanned = banned_id;
    total_user = user_id.size();
    total_ban = banned_id.size();
    
    dfs(0);
    
    return s.size();
}

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.py
def is_match(user_id: str, banned_id: str) -> bool:
    if len(user_id) != len(banned_id):
        return False
    
    for u_char, b_char in zip(user_id, banned_id):
        if b_char == '*':
            continue
            
        if u_char != b_char:
            return False
        
    return True

def solution(user_id, banned_id):
    from collections import deque
    
    num_user = len(user_id)
    num_banned = len(banned_id)
    banned_used = [False for _ in range(num_banned)]
    cur_d = deque()
    answer_s = set()
    
    def dfs(user_idx):
        if user_idx == num_user:
            if all(banned_used):
                answer_s.add(tuple(sorted(cur_d)))
            return
        
        for banned_idx in range(num_banned):
            if banned_used[banned_idx]:
                continue
                
            if is_match(user_id[user_idx], banned_id[banned_idx]):
                banned_used[banned_idx] = True
                cur_d.append(user_id[user_idx])
                dfs(user_idx + 1)
                banned_used[banned_idx] = False
                cur_d.pop()
                
        dfs(user_idx + 1)
        
        return

    dfs(0)
                
    answer = len(answer_s)
    
    return answer